File: src/network/packet/impl/PlayerConnectPacket.php
<?php

namespace pocketcloud\cloud\network\packet\impl;

use pocketcloud\cloud\network\client\ServerClient;
use pocketcloud\cloud\network\packet\CloudboundPacket;
use pocketcloud\cloud\network\packet\CloudPacket;
use pocketcloud\cloud\network\packet\util\PacketData;
use pocketcloud\cloud\player\CloudPlayer;
use pocketcloud\cloud\player\CloudPlayerManager;

final class PlayerConnectPacket extends CloudPacket implements CloudboundPacket {

    public function __construct(private ?CloudPlayer $player = null) {}

    public function handle(ServerClient $client): void {
        if (($server = $client->getServer()) !== null) {
            if (CloudPlayerManager::getInstance()->get($this->player->getName()) === null) {
                if ($server->getTemplate()->getTemplateType()->isServer()) $this->player->setCurrentServer($server);
                else $this->player->setCurrentProxy($server);
                CloudPlayerManager::getInstance()->add($this->player);
            }
        }
    }

    public function encodePayload(PacketData $packetData): void {}

    public function decodePayload(PacketData $packetData): void {
        $packetData->readAllTypeSafe([&$this->player], [fn() => $packetData->readPlayer()]);
    }

    public function getPlayer(): ?CloudPlayer {
        return $this->player;
    }

    public static function create(CloudPlayer $player): self {
        return new self($player);
    }
}

File: src/main/java/de/pocketcloud/cloud/network/packet/impl/PlayerConnectPacket.java
package de.pocketcloud.cloud.network.packet.impl;

import de.pocketcloud.cloud.network.client.ServerClient;
import de.pocketcloud.cloud.network.packet.CloudboundPacket;
import de.pocketcloud.cloud.network.packet.CloudPacket;
import de.pocketcloud.cloud.network.packet.data.PacketData;
import de.pocketcloud.cloud.player.CloudPlayer;
import de.pocketcloud.cloud.player.CloudPlayerManager;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.jetbrains.annotations.NotNull;

@NoArgsConstructor
@Getter
public final class PlayerConnectPacket extends CloudPacket implements CloudboundPacket {

    private CloudPlayer player;

    public PlayerConnectPacket(CloudPlayer player) {
        this.player = player;
    }

    @Override
    public void handle(@NotNull ServerClient client) {
        var server = client.server();
        if (server != null) {
            if (CloudPlayerManager.instance().get(player.name()).isEmpty()) {
                if (server.template().templateType().isServer()) player.setCurrentServer(server);
                else player.setCurrentProxy(server);
                CloudPlayerManager.instance().add(player);
            }
        }
    }

    @Override
    public void encodePayload(PacketData packetData) {}

    @Override
    public void decodePayload(PacketData packetData) {
        this.player = packetData.readPlayer();
    }

    public static PlayerConnectPacket create(CloudPlayer player) {
        return new PlayerConnectPacket(player);
    }
}
